### PR TITLE
feat: add holidays argument to business_day_count

### PR DESCRIPTION
**Pull request by MarcoGorelli**
Business polars: part 3 of 5

Is there a better way to pass the list of holidays on from Python to Rust?

### CODE REVIEW

File: crates/polars-ops/src/series/ops/business.rs [hunk 1]
```diff
@@ -75,10 +94,19 @@ fn business_day_count_impl(
         end_date += 1;
     }
 
+    let holidays_begin = match holidays.binary_search(&start_date) {
+        Ok(x) => x,
+        Err(x) => x,
+    } as i32;
+    let holidays_end = match holidays.binary_search(&end_date) {
```
Comment (ritchie46) [inline crates/polars-ops/src/series/ops/business.rs:101]: Maybe start the second binary search on the location the first has returned.

File: crates/polars-ops/src/series/ops/business.rs [hunk 2]
```diff
@@ -7,14 +8,28 @@ use polars_core::prelude::*;
...
 ) -> PolarsResult<Series> {
     if !week_mask.iter().any(|&x| x) {
         polars_bail!(ComputeError:"`week_mask` must have at least one business day");
     }
+
+    // De-dupe and sort holidays, and exclude non-business days.
+    let mut holidays: Vec<i32> = holidays
+        .iter()
+        .filter(|&x| *unsafe { week_mask.get_unchecked(weekday(*x)) })
+        .cloned()
```
Comment (ritchie46) [inline crates/polars-ops/src/series/ops/business.rs:27]: This should be copied?

File: crates/polars-ops/src/series/ops/business.rs [hunk 3]
```diff
@@ -7,14 +8,28 @@ use polars_core::prelude::*;
...
     start: &Series,
     end: &Series,
     week_mask: [bool; 7],
+    holidays: &[i32],
 ) -> PolarsResult<Series> {
     if !week_mask.iter().any(|&x| x) {
         polars_bail!(ComputeError:"`week_mask` must have at least one business day");
     }
+
+    // De-dupe and sort holidays, and exclude non-business days.
```
Comment (ritchie46) [inline crates/polars-ops/src/series/ops/business.rs:23]: Why would we get duplicates?

File: crates/polars-ops/src/series/ops/business.rs [hunk 4]
```diff
@@ -7,14 +8,28 @@ use polars_core::prelude::*;
...
+    let mut holidays: Vec<i32> = holidays
+        .iter()
+        .filter(|&x| *unsafe { week_mask.get_unchecked(weekday(*x)) })
+        .cloned()
+        .collect::<HashSet<_>>()
+        .into_iter()
+        .collect();
+    holidays.sort_unstable();
```
Comment (ritchie46) [inline crates/polars-ops/src/series/ops/business.rs:31]: Sorting and unique feel very similar operations. I feel that can be in one.

File: crates/polars-ops/src/series/ops/business.rs (hunk 3 above)
Comment (MarcoGorelli) [reply crates/polars-ops/src/series/ops/business.rs]: If a user passes them :) In which case, they'd get incorrect results if Polars doesn't de-dupe for them

File: crates/polars-ops/src/series/ops/business.rs (hunk 4 above)
Comment (MarcoGorelli) [reply crates/polars-ops/src/series/ops/business.rs]: Thanks for your review!

What numpy do is:
1. sort
2. de-dupe + filter out weekends in the same step

I wouldn't know how to combine these further - but note that this operation only needs doing once per `business_day_count` call (it's outside the loop over rows) and that, unless people are extraordinarily lucky, they're probably only passing in a a limited number of holiday days

File: crates/polars-ops/src/series/ops/business.rs [hunk 5]
```diff
@@ -98,6 +116,21 @@ fn business_day_count_impl(
     }
 }
 
+/// Sort and deduplicate holidays and remove holidays that are not business days.
+fn normalise_holidays(holidays: &[i32], week_mask: &[bool; 7]) -> Vec<i32> {
+    let mut holidays: Vec<i32> = holidays.to_vec();
+    holidays.sort_unstable();
+    let mut previous_holiday: Option<i32> = None;
+    holidays.retain(|&x| {
+        if (Some(x) == previous_holiday) || !unsafe { *week_mask.get_unchecked(weekday(x)) } {
```
Comment (ritchie46) [inline crates/polars-ops/src/series/ops/business.rs:125]: Can you add a `// SAFETY` tag explaining why this is sound?

**Review by ritchie46 — APPROVED** [review]
Left one comment on the `unsafe` tag. After that it is good to go. :)

**Comment by MarcoGorelli** [comment]
thanks for your review! 🙏 
shipping it then to move forwards with the next one

**Pull request merged**